### PR TITLE
MTDSA-14297: Unignore Employment TYS

### PR DESCRIPTION
**Pull request by SamuelJacobsHMRC**

### CODE REVIEW

File: app/v1/connectors/IgnoreEmploymentConnector.scala [hunk 1]
```diff
@@ -37,7 +37,7 @@ class IgnoreEmploymentConnector @Inject() (val http: HttpClient, val appConfig:
     import api.connectors.httpparsers.StandardDownstreamHttpParser._
 
     val nino         = request.nino.nino
-    val taxYear      = request.taxYear
+    val taxYear      = request.taxYear.asMtd
```
Comment (mattstephens-hmrc) [inline app/v1/connectors/IgnoreEmploymentConnector.scala:40]: could these follow the pattern used elsewhere with the "import request._" and then inline directly in the string, so the 3 vals aren't needed?

Comment (mattstephens-hmrc) [reply app/v1/connectors/IgnoreEmploymentConnector.scala:40]: (I know you were just adding the .asMtd here, so no worries if it's outside the scope of the PR)

File: app/v1/connectors/UnignoreEmploymentConnector.scala [hunk 2]
```diff
@@ -35,11 +35,15 @@ class UnignoreEmploymentConnector @Inject() (val http: HttpClient, val appConfig
...
     import api.connectors.httpparsers.StandardDownstreamHttpParser._
 
-    val nino         = request.nino.nino
-    val taxYear      = request.taxYear
-    val employmentId = request.employmentId
+    import request._
 
-    delete(IfsUri[Unit](s"income-tax/employments/$nino/$taxYear/ignore/$employmentId"))
+    val downstreamUri = if (taxYear.useTaxYearSpecificApi) {
+      TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/employments/${nino.nino}/ignore/$employmentId")
```
Comment (mattstephens-hmrc) [inline app/v1/connectors/UnignoreEmploymentConnector.scala:41]: nino.toString returns the nino value, so can just do `$nino` here

File: app/v1/connectors/UnignoreEmploymentConnector.scala [hunk 3]
```diff
@@ -21,7 +21,7 @@ import config.AppConfig
 
 import javax.inject.{Inject, Singleton}
 import uk.gov.hmrc.http.{HeaderCarrier, HttpClient}
-import api.connectors.DownstreamUri.IfsUri
+import api.connectors.DownstreamUri.{IfsUri, TaxYearSpecificIfsUri}
 import v1.models.request.ignoreEmployment.IgnoreEmploymentRequest
 
 import scala.concurrent.{ExecutionContext, Future}
```
Comment (mattstephens-hmrc) [inline app/v1/connectors/UnignoreEmploymentConnector.scala:27]: unused import (BaseDownstreamConnector)

File: app/v1/connectors/UnignoreEmploymentConnector.scala [hunk 4]
```diff
@@ -35,11 +35,15 @@ class UnignoreEmploymentConnector @Inject() (val http: HttpClient, val appConfig
 
     import api.connectors.httpparsers.StandardDownstreamHttpParser._
```
Comment (mattstephens-hmrc) [inline app/v1/connectors/UnignoreEmploymentConnector.scala:36]: Not related to your PR, but while you're there, could this be moved up into the import block at the top?

File: app/v1/connectors/UnignoreEmploymentConnector.scala (hunk 3 above)
Comment (mattstephens-hmrc) [reply app/v1/connectors/UnignoreEmploymentConnector.scala:27]: (looks like it's imported twice)

File: app/v1/connectors/UnignoreEmploymentConnector.scala [hunk 5]
```diff
@@ -35,11 +35,15 @@ class UnignoreEmploymentConnector @Inject() (val http: HttpClient, val appConfig
...
+    import request._
 
-    delete(IfsUri[Unit](s"income-tax/employments/$nino/$taxYear/ignore/$employmentId"))
+    val downstreamUri = if (taxYear.useTaxYearSpecificApi) {
+      TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/employments/${nino.nino}/ignore/$employmentId")
+    } else {
+      IfsUri[Unit](s"income-tax/employments/${nino.nino}/${taxYear.asMtd}/ignore/$employmentId")
+    }
+
+    delete(uri = downstreamUri)
```
Comment (mattstephens-hmrc) [inline app/v1/connectors/UnignoreEmploymentConnector.scala:46]: could probably remove the "uri = " as it's a tad redundant

File: app/v1/requestParsers/validators/IgnoreEmploymentValidator.scala [hunk 6]
```diff
@@ -46,7 +46,7 @@ class IgnoreEmploymentValidator @Inject() (implicit currentDateTime: CurrentDate
   private def parameterRuleValidation: IgnoreEmploymentRawData => List[List[MtdError]] = (data: IgnoreEmploymentRawData) => {
     List(
       TaxYearNotSupportedValidation.validate(data.taxYear, appConfig.minimumPermittedTaxYear),
-      if (FeatureSwitches().isTaxYearNotEndedRuleEnabled) TaxYearNotEndedValidation.validate(data.taxYear) else List.empty[MtdError]
+      if (data.temporalValidationEnabled) TaxYearNotEndedValidation.validate(data.taxYear) else List.empty[MtdError]
```
Comment (mattstephens-hmrc) [inline app/v1/requestParsers/validators/IgnoreEmploymentValidator.scala:49]: Can replace `List.empty[MtdError]` with `Nil`

File: app/v1/services/UnignoreEmploymentService.scala [hunk 7]
```diff
@@ -39,15 +39,13 @@ class UnignoreEmploymentService @Inject() (connector: UnignoreEmploymentConnecto
       logContext: EndpointLogContext,
       correlationId: String): Future[Either[ErrorWrapper, ResponseWrapper[Unit]]] = {
 
-    val result = for {
-      desResponseWrapper <- EitherT(connector.unignoreEmployment(request)).leftMap(mapDownstreamErrors(desErrorMap))
-    } yield desResponseWrapper
+    val result = EitherT(connector.unignoreEmployment(request)).leftMap(mapDownstreamErrors(downstreamErrorMap))
```
Comment (mattstephens-hmrc) [inline app/v1/services/UnignoreEmploymentService.scala:42]: can simplify & remove the EitherT

File: app/v1/services/UnignoreEmploymentService.scala [hunk 8]
```diff
@@ -39,15 +39,13 @@ class UnignoreEmploymentService @Inject() (connector: UnignoreEmploymentConnecto
...
 
     result.value
   }
 
-  private def desErrorMap: Map[String, MtdError] =
-    Map(
+  private def downstreamErrorMap: Map[String, MtdError] = {
```
Comment (mattstephens-hmrc) [inline app/v1/services/UnignoreEmploymentService.scala:47]: should be a val

File: app/v1/connectors/IgnoreEmploymentConnector.scala (hunk 1 above)
Comment (SamuelJacobsHMRC) [reply app/v1/connectors/IgnoreEmploymentConnector.scala:40]: Jeremy was working on the ignore endpoint and its actually been merged now so this should be all sorted. I will merge in main and resolve the conflicts which should resolve everything related to the Ignore endpoint

File: app/v1/connectors/UnignoreEmploymentConnector.scala (hunk 3 above)
Comment (SamuelJacobsHMRC) [reply app/v1/connectors/UnignoreEmploymentConnector.scala:27]: Oh strange, will update

File: it/v1/endpoints/UnignoreEmploymentControllerISpec.scala [hunk 9]
```diff
@@ -123,18 +95,19 @@ class UnignoreEmploymentControllerISpec extends IntegrationBaseSpec {
...
           ("AA123456A", getCurrentTaxYear, "78d9f015-a8b4-47a8-8bbc-c253a1e8057e", BAD_REQUEST, RuleTaxYearNotEndedError, None)
         )
+
         input.foreach(args => (validationErrorTest _).tupled(args))
       }
 
-      "ifs service error" when {
-        def serviceErrorTest(ifsStatus: Int, ifsCode: String, expectedStatus: Int, expectedBody: MtdError): Unit = {
-          s"ifs returns an $ifsCode error and status $ifsStatus" in new Test {
+      "service error" when {
```
Comment (mattstephens-hmrc) [inline it/v1/endpoints/UnignoreEmploymentControllerISpec.scala:102]: I think one of these nested test levels can be removed, here and for "validation error".

So you'd have a top-level `"return a validation error according to spec" when {`
and a top-level `"return a service error according to spec" when {`

File: it/v1/endpoints/UnignoreEmploymentControllerISpec.scala [hunk 10]
```diff
@@ -162,9 +127,78 @@ class UnignoreEmploymentControllerISpec extends IntegrationBaseSpec {
           (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, StandardDownstreamError),
           (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, StandardDownstreamError)
         )
-        input.foreach(args => (serviceErrorTest _).tupled(args))
+
+        val extraTysErrors = Seq(
```
Comment (mattstephens-hmrc) [inline it/v1/endpoints/UnignoreEmploymentControllerISpec.scala:131]: This is a really minor thing, but it's preferable to use List() when creating a collection. Seq for type definitions, List or Vector etc (i.e. specify the implementation you want) when creating them.

Comment (mattstephens-hmrc) [reply it/v1/endpoints/UnignoreEmploymentControllerISpec.scala:131]: (it's a minor example, but a major coding principle :slightly_smiling_face:)

File: test/v1/connectors/UnignoreEmploymentConnectorSpec.scala [hunk 11]
```diff
@@ -17,59 +17,58 @@
...
-  )
+        willDelete(
+          url = s"$baseUrl/income-tax/employments/$nino/2019-20/ignore/$employmentId"
+        ).returns(Future.successful(outcome))
 
-  class Test extends MockHttpClient with MockAppConfig {
+        await(connector.unignoreEmployment(request)) shouldBe outcome
+      }
+    }
+    "return the expected response for a TYS request" when {
```
Comment (mattstephens-hmrc) [inline test/v1/connectors/UnignoreEmploymentConnectorSpec.scala:41]: can these be "spaced out" a bit?

File: test/v1/connectors/UnignoreEmploymentConnectorSpec.scala [hunk 12]
```diff
@@ -17,59 +17,58 @@
...
-    nino = Nino(nino),
-    taxYear = taxYear,
-    employmentId = employmentId
-  )
+        willDelete(
+          url = s"$baseUrl/income-tax/employments/$nino/2019-20/ignore/$employmentId"
+        ).returns(Future.successful(outcome))
 
-  class Test extends MockHttpClient with MockAppConfig {
+        await(connector.unignoreEmployment(request)) shouldBe outcome
```
Comment (mattstephens-hmrc) [inline test/v1/connectors/UnignoreEmploymentConnectorSpec.scala:38]: can the result be separated out, i.e.:
```
        val result: DownstreamOutcome[Unit] = await(connector.unignoreEmployment(request))
        result shouldBe expectedOutcome
```
(can also rename outcome to expectedOutcome),

& same convention elsewhere

Comment (mattstephens-hmrc) [reply test/v1/connectors/UnignoreEmploymentConnectorSpec.scala:38]: adding the result type also makes the test more specific which is probably a good thing :)

File: test/v1/services/UnignoreEmploymentServiceSpec.scala [hunk 13]
```diff
@@ -17,38 +17,20 @@
...
-    val service: UnignoreEmploymentService = new UnignoreEmploymentService(
-      connector = mockUnignoreEmploymentConnector
-    )
-
-  }
-
-  "UnignoreEmploymentService" when {
-    "unignoreEmployment" should {
+  "UnignoreEmploymentService" should {
+    "unignoreEmployment" must {
```
[the remark this answers is not in the document]
Comment (mattstephens-hmrc) [reply test/v1/services/UnignoreEmploymentServiceSpec.scala:33]: actually the top-level `"UnignoreEmploymentService" should {}` can be deleted.  Doesn't need the service class name as that's in the Spec class name; so just the `unignoreEmployment()` function name is sufficient.

File: test/v1/services/UnignoreEmploymentServiceSpec.scala [hunk 14]
```diff
@@ -60,18 +42,17 @@ class UnignoreEmploymentServiceSpec extends ServiceSpec {
...
             MockUnignoreEmploymentConnector
               .unignoreEmployment(request)
-              .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(desErrorCode))))))
+              .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(downstreamErrorCode))))))
 
             await(service.unignoreEmployment(request)) shouldBe Left(ErrorWrapper(correlationId, error))
           }
 
-        val input = Seq(
+        val errors = Seq(
```
Comment (mattstephens-hmrc) [inline test/v1/services/UnignoreEmploymentServiceSpec.scala:55]: same suggestion about List

**Review by mattstephens-hmrc — APPROVED** [review]

**Pull request merged**